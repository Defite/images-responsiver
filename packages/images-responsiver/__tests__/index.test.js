'use strict';

const prettier = require('prettier');
const imagesResponsiver = require('../index');

function cleanHtml(html) {
  return prettier.format(html, { parser: 'html' });
}

describe('no image', () => {
  test('keeps intact HTML without image', () => {
    const content = `<!DOCTYPE html><html><body><p>Hello</p></body></html>`;
    const transformed = imagesResponsiver(content);
    expect(cleanHtml(transformed)).toMatchSnapshot();
  });
});

describe("image that can't be transformed", () => {
  test('do nothing on SVG image', () => {
    const content = `<!DOCTYPE html><html><body><img src="test.svg"></body></html>`;
    const transformed = imagesResponsiver(content);
    expect(cleanHtml(transformed)).toMatchSnapshot();
  });

  test('do nothing on image without a src', () => {
    const content = `<!DOCTYPE html><html><body><img alt="not really an image"></body></html>`;
    const transformed = imagesResponsiver(content);
    const expected = content;
    expect(cleanHtml(transformed)).toMatchSnapshot();
  });

  test('do nothing on image with already a srcset', () => {
    const content = `<!DOCTYPE html><html><body><img src="test.png" srcset="test.png 320w"></body></html>`;
    const transformed = imagesResponsiver(content);
    const expected = content;
    expect(cleanHtml(transformed)).toMatchSnapshot();
  });
});

describe('image without options', () => {
  test('simple image', () => {
    const content = `<!DOCTYPE html><html><body><img src="test.png"></body></html>`;
    const transformed = imagesResponsiver(content);
    expect(cleanHtml(transformed)).toMatchSnapshot();
  });

  test('width attribute superior to the fallback', () => {
    const content = `<!DOCTYPE html><body>
      <img src="test.png" width="789"></body>`;
    const transformed = imagesResponsiver(content);
    expect(cleanHtml(transformed)).toMatchSnapshot();
  });

  test('width attribute inferior to the fallback', () => {
    const content = `<!DOCTYPE html><body>
      <img src="test.png" width="543"></body>`;
    const transformed = imagesResponsiver(content);
    expect(cleanHtml(transformed)).toMatchSnapshot();
  });

  test('width attribute inferior to the minWidth', () => {
    const content = `<!DOCTYPE html><body>
      <img src="test.png" width="200"></body>`;
    const transformed = imagesResponsiver(content);
    expect(cleanHtml(transformed)).toMatchSnapshot();
  });

  test('2 images', () => {
    const content = `<!DOCTYPE html><body>
      <img src="test1.png" width="350" /><img src="test2.png" width="400" /></body>`;
    const transformed = imagesResponsiver(content);

    expect(cleanHtml(transformed)).toMatchSnapshot();
  });
});

describe('image with options', () => {
  test('simple image', () => {
    const content = `<!DOCTYPE html><html><body><img src="test.png"></body></html>`;
    const transformed = imagesResponsiver(content, {
      default: {
        minWidth: 120,
        maxWidth: 320,
        steps: 3,
      },
    });
    expect(cleanHtml(transformed)).toMatchSnapshot();
  });

  test('width attribute superior to the fallback', () => {
    const content = `<!DOCTYPE html><body>
      <img src="test.png" width="789"></body>`;
    const transformed = imagesResponsiver(content, {
      default: {
        fallbackWidth: 480,
      },
    });
    expect(cleanHtml(transformed)).toMatchSnapshot();
  });

  test('width attribute inferior to the fallback', () => {
    const content = `<!DOCTYPE html><body>
      <img src="test.png" width="420"></body>`;
    const transformed = imagesResponsiver(content, {
      default: {
        fallbackWidth: 480,
      },
    });
    expect(cleanHtml(transformed)).toMatchSnapshot();
  });
});

describe('advanced options', () => {
  test('selector', () => {
    const content = `<!DOCTYPE html><body>
      <img src="test1.png" /><img src="test2.png" class="notransform" /></body>`;
    const transformed = imagesResponsiver(content, {
      default: {
        selector: 'img:not(.notransform)',
        fallbackWidth: 480,
      },
    });

    expect(cleanHtml(transformed)).toMatchSnapshot();
  });

  test('resizedImageUrl', () => {
    const content = `<!DOCTYPE html><body>
      <img src="test.png" width="420"></body>`;
    const transformed = imagesResponsiver(content, {
      default: {
        resizedImageUrl: (src, width) => `${src}?w=${width}`,
        fallbackWidth: 480,
      },
    });

    expect(cleanHtml(transformed)).toMatchSnapshot();
  });

  test('runBefore', () => {
    const content = `<!DOCTYPE html><body>
      <img src="test.png"></body>`;
    const transformed = imagesResponsiver(content, {
      default: {
        runBefore: (image, document) => {
          image.setAttribute('src', 'another-test.png');
        },
      },
    });

    expect(cleanHtml(transformed)).toMatchSnapshot();
  });

  test('runAfter', () => {
    const content = `<!DOCTYPE html><body>
      <img src="test.png"></body>`;
    const transformed = imagesResponsiver(content, {
      default: {
        runAfter: (image, document) => {
          image.dataset.after = 'hello!';
        },
      },
    });

    expect(cleanHtml(transformed)).toMatchSnapshot();
  });
});
