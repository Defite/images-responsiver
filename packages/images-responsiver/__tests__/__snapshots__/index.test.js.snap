// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`advanced options resizedImageUrl 1`] = `
"<!DOCTYPE html>
<html>
  <body>
    <img
      src=\\"test.png?w=420\\"
      width=\\"420\\"
      srcset=\\"
        test.png?w=320 320w,
        test.png?w=345 345w,
        test.png?w=370 370w,
        test.png?w=395 395w,
        test.png?w=420 420w
      \\"
      sizes=\\"100vw\\"
      data-pristine=\\"test.png\\"
    />
  </body>
</html>
"
`;

exports[`advanced options selector 1`] = `
"<!DOCTYPE html>
<html>
  <body>
    <img
      src=\\"test1-480.png\\"
      srcset=\\"
        test1-320.png   320w,
        test1-880.png   880w,
        test1-1440.png 1440w,
        test1-2000.png 2000w,
        test1-2560.png 2560w
      \\"
      sizes=\\"100vw\\"
      data-pristine=\\"test1.png\\"
    /><img src=\\"test2.png\\" class=\\"notransform\\" />
  </body>
</html>
"
`;

exports[`image that can't be transformed do nothing on SVG image 1`] = `
"<!DOCTYPE html>
<html>
  <body>
    <img src=\\"test.svg\\" />
  </body>
</html>
"
`;

exports[`image that can't be transformed do nothing on image with already a srcset 1`] = `
"<!DOCTYPE html>
<html>
  <body>
    <img src=\\"test.png\\" srcset=\\"test.png 320w\\" />
  </body>
</html>
"
`;

exports[`image that can't be transformed do nothing on image without a src 1`] = `
"<!DOCTYPE html>
<html>
  <body>
    <img alt=\\"not really an image\\" />
  </body>
</html>
"
`;

exports[`image with options simple image 1`] = `
"<!DOCTYPE html>
<html>
  <body>
    <img
      src=\\"test-640.png\\"
      srcset=\\"test-120.png 120w, test-220.png 220w, test-320.png 320w\\"
      sizes=\\"100vw\\"
      data-pristine=\\"test.png\\"
    />
  </body>
</html>
"
`;

exports[`image with options width attribute inferior to the fallback 1`] = `
"<!DOCTYPE html>
<html>
  <body>
    <img
      src=\\"test-420.png\\"
      width=\\"420\\"
      srcset=\\"
        test-320.png 320w,
        test-345.png 345w,
        test-370.png 370w,
        test-395.png 395w,
        test-420.png 420w
      \\"
      sizes=\\"100vw\\"
      data-pristine=\\"test.png\\"
    />
  </body>
</html>
"
`;

exports[`image with options width attribute superior to the fallback 1`] = `
"<!DOCTYPE html>
<html>
  <body>
    <img
      src=\\"test-480.png\\"
      width=\\"789\\"
      srcset=\\"
        test-320.png 320w,
        test-438.png 438w,
        test-555.png 555w,
        test-672.png 672w,
        test-789.png 789w
      \\"
      sizes=\\"100vw\\"
      data-pristine=\\"test.png\\"
    />
  </body>
</html>
"
`;

exports[`image without options 2 images 1`] = `
"<!DOCTYPE html>
<html>
  <body>
    <img
      src=\\"test1-350.png\\"
      width=\\"350\\"
      srcset=\\"
        test1-320.png 320w,
        test1-328.png 328w,
        test1-335.png 335w,
        test1-343.png 343w,
        test1-350.png 350w
      \\"
      sizes=\\"100vw\\"
      data-pristine=\\"test1.png\\"
    /><img
      src=\\"test2-400.png\\"
      width=\\"400\\"
      srcset=\\"
        test2-320.png 320w,
        test2-340.png 340w,
        test2-360.png 360w,
        test2-380.png 380w,
        test2-400.png 400w
      \\"
      sizes=\\"100vw\\"
      data-pristine=\\"test2.png\\"
    />
  </body>
</html>
"
`;

exports[`image without options simple image 1`] = `
"<!DOCTYPE html>
<html>
  <body>
    <img
      src=\\"test-640.png\\"
      srcset=\\"
        test-320.png   320w,
        test-880.png   880w,
        test-1440.png 1440w,
        test-2000.png 2000w,
        test-2560.png 2560w
      \\"
      sizes=\\"100vw\\"
      data-pristine=\\"test.png\\"
    />
  </body>
</html>
"
`;

exports[`image without options width attribute inferior to the fallback 1`] = `
"<!DOCTYPE html>
<html>
  <body>
    <img
      src=\\"test-543.png\\"
      width=\\"543\\"
      srcset=\\"
        test-320.png 320w,
        test-376.png 376w,
        test-432.png 432w,
        test-488.png 488w,
        test-543.png 543w
      \\"
      sizes=\\"100vw\\"
      data-pristine=\\"test.png\\"
    />
  </body>
</html>
"
`;

exports[`image without options width attribute inferior to the minWidth 1`] = `
"<!DOCTYPE html>
<html>
  <body>
    <img
      src=\\"test-200.png\\"
      width=\\"200\\"
      srcset=\\"
        test-320.png 320w,
        test-290.png 290w,
        test-260.png 260w,
        test-230.png 230w,
        test-200.png 200w
      \\"
      sizes=\\"100vw\\"
      data-pristine=\\"test.png\\"
    />
  </body>
</html>
"
`;

exports[`image without options width attribute superior to the fallback 1`] = `
"<!DOCTYPE html>
<html>
  <body>
    <img
      src=\\"test-640.png\\"
      width=\\"789\\"
      srcset=\\"
        test-320.png 320w,
        test-438.png 438w,
        test-555.png 555w,
        test-672.png 672w,
        test-789.png 789w
      \\"
      sizes=\\"100vw\\"
      data-pristine=\\"test.png\\"
    />
  </body>
</html>
"
`;

exports[`no image keeps intact HTML without image 1`] = `
"<!DOCTYPE html>
<html>
  <body>
    <p>Hello</p>
  </body>
</html>
"
`;
